src/main/java/com/hazelcast/azure/AzureDiscoveryStrategy.java

```diff
@@ -88,63 +81,103 @@ public void start() {
-                DiscoveryNode node = buildDiscoveredNode(faultDomainId, vm, port);
-
-                if (node != null) {
-                    nodes.add(node);
-                }
-            }
             LOGGER.info("Azure Discovery SPI Discovered " + nodes.size() + " nodes");
             return nodes;
         } catch (Exception e) {
             LOGGER.finest("Failed to discover nodes with Azure SPI", e);
             return null;
         }
+    }
 
+    private SimpleDiscoveryNode createDiscoveryNode(int port, VirtualMachineScaleSetVM vm,
+                                                    VirtualMachineScaleSetNetworkInterface networkInterface)
+            throws UnknownHostException {
+        String privateIP = networkInterface.primaryPrivateIP();
+        if (getLocalHostAddress() != null && privateIP.equals(getLocalHostAddress())) {
+            Integer faultDomainId = vm.instanceView().platformFaultDomain();
+            updateVirtualMachineMetaData(faultDomainId);
+        }
+        Address privateAddress = new Address(privateIP, port);
+        return new SimpleDiscoveryNode(privateAddress);
     }
 
-    @Override
-    public void destroy() {
-        // no native resources were allocated so nothing to do here
+    private List<DiscoveryNode> discoverScaleSetVMs(String resourceGroup, String clusterId)
+            throws UnknownHostException {
+        PagedList<VirtualMachineScaleSet> scaleSets = computeManager.virtualMachineScaleSets()
+                .listByResourceGroup(resourceGroup);
+        ArrayList<DiscoveryNode> nodes = new ArrayList<DiscoveryNode>();
+
+        for (VirtualMachineScaleSet scaleSet : scaleSets) {
+            Map<String, String> tags = scaleSet.tags();
+            // a tag is required with the hazelcast clusterid
+            // and the value should be the port number
+            if (tags.get(clusterId) == null) {
+                continue;
+            }
+            int port = Integer.parseInt(tags.get(clusterId));
+
+            PagedList<VirtualMachineScaleSetVM> vms = scaleSet.virtualMachines().list();
+
+            for (VirtualMachineScaleSetVM vm : vms) {
+                if (!PowerState.RUNNING.equals(vm.powerState())) {
+                    continue;
+                }
+
+                String primaryNetworkInterfaceId = vm.primaryNetworkInterfaceId();
+                if (primaryNetworkInterfaceId != null) {
+                    VirtualMachineScaleSetNetworkInterface networkInterface =
+                            vm.getNetworkInterface(primaryNetworkInterfaceId);
+                    nodes.add(createDiscoveryNode(port, vm, networkInterface));
+                } else {
+                    PagedList<VirtualMachineScaleSetNetworkInterface> networkInterfaces = vm.listNetworkInterfaces();
+                    if (networkInterfaces.size() > 0) {
+                        VirtualMachineScaleSetNetworkInterface networkInterface = networkInterfaces.get(0);
```

Comment:
Why the first one?

Reply:
It's a bit tricky if a scale set has multiple network interface then one of them will be primary and then we can retrieve the primary interface.  We do that in the `if` block. If a scale set has only one network interface then `VirtualMachineScaleSetVM.primaryNetworkInterfaceId()` returns `null` and then we have to iterate through the network interfaces and pick the first one. Hope this makes sense.